### PR TITLE
Add typescript-inversify code generator

### PR DESCRIPTION
**Pull request by gualtierim**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [x] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
We added a code generator called `typescript-inversify` to support InversifyJS: a powerful and lightweight inversion of control container for JavaScript & Node.js apps powered by TypeScript. Inside the README is explained how to integrate services generated in an existing project.

### CODE REVIEW

File: bin/security/typescript-inversify.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
Comment (macjohnny) [inline bin/security/typescript-inversify.sh]: please add a .cmd version of this file for windows-users

File: bin/typescript-inversify-petstore.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
Comment (macjohnny) [inline bin/typescript-inversify-petstore.sh]: please add a .cmd version of this file for windows-users

File: modules/swagger-codegen/src/main/resources/typescript-inversify/HttpClient.mustache [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import IHttpClient from "./IHttpClient";
+import * as Rx from "rx";
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/HttpClient.mustache]: wouldn't it be better to 
```
import { Observable } from "rxjs/Observable";
```
?

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+{{/required}}
+{{/allParams}}
+{{#hasQueryParams}}
+        let queryParameters: string[] = [];
+{{#queryParams}}
+        {{#isListContainer}}
+        if ({{paramName}}) {
+        {{#isCollectionFormatMulti}}
+            {{paramName}}.forEach((element) => {
+                queryParameters.push("{{paramName}}="+String({{paramName}}));
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: you need to `encodeURIComponent()`, 
```
queryParameters.push("{{paramName}}="+encodeURIComponent(String({{paramName}})));
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        {{/isCollectionFormatMulti}}
+        {{^isCollectionFormatMulti}}
+            queryParameters.push("{{paramName}}="+{{paramName}}.join(COLLECTION_FORMATS['{{collectionFormat}}']));
+        {{/isCollectionFormatMulti}}
+        }
+        {{/isListContainer}}
+        {{^isListContainer}}
+        if ({{paramName}} !== undefined) {
+        {{#isDateTime}}
+           queryParameters.push("{{paramName}}="+<any>{{paramName}}.toISOString());
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: you need to `encodeURIComponent()`, 
```
queryParameters.push("{{paramName}}="+encodeURIComponent(<any>{{paramName}}.toISOString()));
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        {{/isDateTime}}
+        {{^isDateTime}}
+            queryParameters.push("{{paramName}}="+String({{paramName}}));
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: you need to `encodeURIComponent()`, 
```
queryParameters.push("{{paramName}}="+encodeURIComponent(String({{paramName}})));
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+{{#queryParams}}
+        {{#isListContainer}}
+        if ({{paramName}}) {
+        {{#isCollectionFormatMulti}}
+            {{paramName}}.forEach((element) => {
+                queryParameters.push("{{paramName}}="+String({{paramName}}));
+            })
+        {{/isCollectionFormatMulti}}
+        {{^isCollectionFormatMulti}}
+            queryParameters.push("{{paramName}}="+{{paramName}}.join(COLLECTION_FORMATS['{{collectionFormat}}']));
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: you need to `encodeURIComponent()`, 
```
queryParameters.push("{{paramName}}="+encodeURIComponent({{paramName}}.join(COLLECTION_FORMATS['{{collectionFormat}}'])));
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        // authentication ({{name}}) required
+{{#isApiKey}}
+{{#isKeyInHeader}}
+        if (this.APIConfiguration.apiKeys["{{keyParamName}}"]) {
+            headers['{{keyParamName}}'] = this.APIConfiguration.apiKeys["{{keyParamName}}"];
+        }
+{{/isKeyInHeader}}
+{{#isKeyInQuery}}
+        if (this.APIConfiguration.apiKeys["{{keyParamName}}"]) {
+            queryParameters.set('{{keyParamName}}', this.APIConfiguration.apiKeys["{{keyParamName}}"]);
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: how can `queryParameters.set()` work on a `string[]` ?
should be
```
queryParameters.push("{{keyParamName}}="+encodeURIComponent(this.APIConfiguration.apiKeys["{{keyParamName}}"]));
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        {{^isListContainer}}
+        if ({{paramName}} !== undefined) {
+            formData.append('{{baseName}}', <any>{{paramName}});
+        }
+        {{/isListContainer}}
+{{/formParams}}
+
+{{/hasFormParams}}
+
+        return this.httpClient.{{httpMethod}}(`${this.basePath}{{{path}}}{{#hasQueryParams}}?${encodeURIComponent(queryParameters.join('&'))}{{/hasQueryParams}}`{{#bodyParam}}, {{paramName}} {{/bodyParam}}{{#hasFormParams}}, body{{/hasFormParams}}, headers)
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: `encodeURIComponent(queryParameters.join('&'))`` should be `queryParameters.join('&')`, otherwise the `=` between the parameter name and the value will be converted to `%3D`

File: modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    configuration: Configuration;
+    [others: string]: any;
+
+{{#operation}}
+    /**
+    * {{summary}}
+    * {{notes}}
+    {{#allParams}}* @param {{paramName}} {{description}}
+    {{/allParams}}*/
+    {{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}extraHttpRequestParams?: any): Observable<{{#returnType}}{{{returnType}}}{{/returnType}}{{^returnType}}{}{{/returnType}}>;
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache]: the signature should correspond to the api service, i.e. should be
```
{{nickname}}({{#allParams}}{{paramName}}{{^required}}?{{/required}}: {{{dataType}}}, {{/allParams}}headers: Dictionary<string> = {}): Promise<{{#returnType}}{{{returnType}}}{{#isResponseTypeFile}}|undefined{{/isResponseTypeFile}}{{/returnType}}{{^returnType}}any{{/returnType}}>
```

File: modules/swagger-codegen/src/main/resources/typescript-inversify/package.mustache [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  "main": "dist/index.js",
+  "typings": "dist/index.d.ts",
+  "scripts": {
+    "build": "tsc --outDir dist/",
+    "postinstall": "npm run build"
+  },
+  "dependencies": {
+    "inversify": "^4.3.0",
+    "lodash": "^4.17.5",
+    "rx": "~4.1.0",
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/package.mustache]: wouldn't you want to have `rxjs`: "~5.5.2" ?

File: modules/swagger-codegen/src/test/java/io/swagger/codegen/typescript/typescriptinversify/TypeScriptInversifyClientOptionsTest.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package io.swagger.codegen.typescript.typescriptinversify;
+
+import io.swagger.codegen.AbstractOptionsTest;
+import io.swagger.codegen.CodegenConfig;
+import io.swagger.codegen.languages.TypeScriptInversifyClientCodegen;
+import io.swagger.codegen.options.TypeScriptInversifyClientOptionsProvider;
+import mockit.Expectations;
+import mockit.Tested;
+
+public class TypeScriptInversifyClientOptionsTest extends AbstractOptionsTest {
```
Comment (macjohnny) [inline modules/swagger-codegen/src/test/java/io/swagger/codegen/typescript/typescriptinversify/TypeScriptInversifyClientOptionsTest.java]: very nice

**Comment by macjohnny** [comment]
cc @TiFu @taxpon @sebastianhaas @kenisteward @Vrolijkx

**Comment by gualtierim** [comment]
Hi, thanks for all yours suggestions. I should have solved all the problems listed.

File: samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import * as Rx from "rx";
+import HttpResponse from "./HttpResponse";
+import { Dictionary } from "lodash";
+
+interface IHttpClient {
+    get(url:string, headers?:Dictionary<string>):Rx.Observable<HttpResponse>
```
Comment (macjohnny) [inline samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts]: do you really want to use a lodash dictionary? why not use a plain JavaScript object?

File: samples/client/petstore/typescript-inversify/api/pet.service.ts [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        if (this.APIConfiguration.accessToken) {
+            let accessToken = typeof this.APIConfiguration.accessToken === 'function'
+                ? this.APIConfiguration.accessToken()
+                : this.APIConfiguration.accessToken;
+            headers['Authorization'] = 'Bearer ' + accessToken;
+        }
+        headers['Accept'] = 'application/xml';
+
+
+        return this.httpClient.get(`${this.basePath}/pet/findByStatus?${encodeURIComponent(queryParameters.join('&'))}`, headers)
```
Comment (macjohnny) [inline samples/client/petstore/typescript-inversify/api/pet.service.ts]: it looks like you need to re-generate the file, since the `encodeURIComponent` should be removed here

File: samples/client/petstore/typescript-inversify/api/store.service.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    public placeOrder(body: Order, headers: Dictionary<string> = {}): Promise<Order> {
+        if (!body){
+            throw new Error('Required parameter body was null or undefined when calling placeOrder.');
+        }
+
+        headers['Accept'] = 'application/xml';
+        headers['Content-Type'] = 'application/json';
+
+
+        return this.httpClient.post(`${this.basePath}/store/order`, body , headers)
```
Comment (macjohnny) [inline samples/client/petstore/typescript-inversify/api/store.service.ts]: are you sure you want to convert all response observables to promises?
returning them as observables would make it convenient to chain with `RxJS`, similar to the Angular `HttpClient`.
As a drawback, you always need to subscribe to the API call observable, otherwise the request is not performed.

File: samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts (hunk 13 above)
Comment (gualtierim) [reply samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts]: The lodash dictionary restrict the type of values and not permit nested Object. Also Angular's HttpClient allows only string as value ([Header]: https://angular.io/api/http/Headers)

Comment (macjohnny) [reply samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts]: you can restrict the value of an object as follows:
```
let myObject: {[key: string]: string}
```

**Comment by macjohnny** [comment]
in general, I would not use lodash here, as it is not needed for any functionality and it adds a dependency. moreover, not everyone wants to use lodash.

File: samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts (hunk 13 above)
Comment (gualtierim) [reply samples/client/petstore-security-test/typescript-inversify/IHttpClient.ts]: You have reason, i'm removing lodash.

File: samples/client/petstore/typescript-inversify/api/store.service.ts (hunk 15 above)
Comment (gualtierim) [reply samples/client/petstore/typescript-inversify/api/store.service.ts]: On this subject I was undecided, because the observable returned by the HttpClient complete immediately and returns only a value (as every Promise). But you have reason that returing a observable would make convenient to chain with `RxJS`, but you can do it with the operator fromPromise if necessary... I'm very undecided about this.

Comment (macjohnny) [reply samples/client/petstore/typescript-inversify/api/store.service.ts]: I would suggest to adapt the interface of the Angular HttpClient with Observables. 
Anyone else got an opinion on this?

File: modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+{{#operations}}
+
+{{#description}}
+    /**
+    * {{&description}}
+    */
+{{/description}}
+export interface {{classname}}Interface {
+    [others: string]: any;
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache]: is this needed?

File: modules/swagger-codegen/src/main/resources/typescript-inversify/configuration.mustache [hunk 17]
```diff
@@ -1,79 +0,0 @@
-export interface ConfigurationParameters {
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/configuration.mustache]: @gualtierim why did you remove the configuration template?

File: modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache (hunk 16 above)
Comment (gualtierim) [reply modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache]: If you set the additional parameters `withInterfaces`, the `apiInterface` should be necessary.

Comment (macjohnny) [reply modules/swagger-codegen/src/main/resources/typescript-inversify/apiInterface.mustache]: I was referring to `[others: string]: any;`

File: samples/client/petstore/typescript-inversify/api/store.service.ts (hunk 15 above)
Comment (gualtierim) [reply samples/client/petstore/typescript-inversify/api/store.service.ts]: I would add a configuration variable that allow the user to decide if use Promise or Observable. What about this solution?

Comment (gualtierim) [reply samples/client/petstore/typescript-inversify/api/store.service.ts]: @macjohnny i've added the `usePromise` config's variable.

**Comment by macjohnny** [comment]
@gualtierim your code looks great.
One thing that I think would be useful is to be able to access the response headers, etc, too. Maybe you could implement this similar to the typescript-angular generator?

**Comment by gualtierim** [comment]
@macjohnny i have added the possibility to receive all the httpResponse. I tried to implement it as in typescript-angular generator.

File: samples/client/petstore/typescript-inversify/api/pet.service.ts [hunk 18]
```diff
@@ -36,15 +37,15 @@ export class PetService {
...
-
     /**
      * Add a new pet to the store
      * 
      * @param body Pet object that needs to be added to the store
      
      */
-
-    public addPet(body: Pet, headers: Headers = {}): Observable<any> {
+    public addPet(body: Pet, observe?: 'body', headers?: Headers): Observable<any>;
```
Comment (macjohnny) [inline samples/client/petstore/typescript-inversify/api/pet.service.ts]: @gualtierim can you find out why the response is `any`? it should be `Pet`, shouldn't it?

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache [hunk 19]
```diff
@@ -168,9 +169,10 @@ export class {{classname}} {
 {{/formParams}}
 
 {{/hasFormParams}}
-
-        return this.httpClient.{{httpMethod}}(`${this.basePath}{{{path}}}{{#hasQueryParams}}?${queryParameters.join('&')}{{/hasQueryParams}}`{{#bodyParam}}, {{paramName}} {{/bodyParam}}{{#hasFormParams}}, body{{/hasFormParams}}, headers)
-                    .map(httpResponse => <{{#returnType}}{{{returnType}}}{{#isResponseTypeFile}}|undefined{{/isResponseTypeFile}}{{/returnType}}{{^returnType}}any{{/returnType}}>(httpResponse.response)){{#usePromise}}.toPromise(){{/usePromise}};
+        const response: Observable<HttpResponse<{{#returnType}}{{{returnType}}}{{#isResponseTypeFile}}|undefined{{/isResponseTypeFile}}{{/returnType}}{{^returnType}}any{{/returnType}}>> = this.httpClient.{{httpMethod}}(`${this.basePath}{{{path}}}{{#hasQueryParams}}?${queryParameters.join('&')}{{/hasQueryParams}}`{{#bodyParam}}, {{paramName}} {{/bodyParam}}{{#hasFormParams}}, body{{/hasFormParams}}, headers);
+        if(observe == 'body')
```
Comment (macjohnny) [inline modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: please wrap the `if` block with curly braces `{` and `}`

**Comment by macjohnny** [comment]
@gualtierim looks great!

File: samples/client/petstore/typescript-inversify/api/pet.service.ts (hunk 18 above)
Comment (macjohnny) [reply samples/client/petstore/typescript-inversify/api/pet.service.ts]: ah, it's because the `addPet` endpoint does not return anything, so it is ok

File: modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache (hunk 19 above)
Comment (gualtierim) [reply modules/swagger-codegen/src/main/resources/typescript-inversify/api.service.mustache]: I done it

**Comment by gualtierim** [comment]
@macjohnny thanks to your support

**Comment by macjohnny** [comment]
@gualtierim did you test the generated code in an actual project?

**Comment by gualtierim** [comment]
Yes, i have tested it inside a project and it seems to work.

**Comment by wing328** [comment]
I'll take another look and merge this weekend. Thanks for the new generator!

**Comment by wing328** [comment]
It seems like there's an issue with as I got the following errors when running "mvn clean test " locally:
```
Failed tests: 
  SpringOptionsTest>AbstractOptionsTest.checkOptionsHelp:39 » NullPointer
  SpringOptionsTest>AbstractOptionsTest.checkOptionsProcessing:28 » NullPointer
```
(which seems totally unrelated)
I've removed it via https://github.com/swagger-api/swagger-codegen/pull/7983 for the time being. Please take a look when you've time.
